### ISSUE TITLE
[Feature Request] Create playlists by folders

### ISSUE THREAD

**Issue by NicoHood** [nims11/IPod-Shuffle-4g#13]
It would be nice to create a playlist for each folder. Or at least somehow set a marker.

Lets say you have:
```
Eminem
 -album1
 -album2
 -album3
David Guetta
 -album1
 -album2
```

It would be nice to have per interpret playlists. But maybe one also only want a single album as playlist. So maybe placing a `.playlist` inside the folder just creates a playlist for the folder and all its subfolders.

Another question:
Is it also possible to enable voiceover for playlist names (havent tried voiceover yet)? The best would be to use voiceover for playlists but not for the songs as it takes too much data. But then you could have tons of playlists which are not names just "playlist 1,2,3...".

**Comment by nims11** [comment]
I too had this idea in mind to have an option to create playlists based on artists/directory, will think on the design and roll out on future updates.

When you enable voiceover, it also works for playlists. adding selective voiceover won't be an issue.

**Comment by NicoHood** [comment]
This would be two really cool features!

Because I do not want to mess with m3u playlists. I just want to select music and put it onto the ipod. and automatic playlists would be nice. you can rather autogenerate for all folder, for only the subfolders or the folders in root. And optional via special `.playlist` selection.

Enabling voiceover for playlists only would be very nice, as it does not take that much data but makes the usage better when switching playlists i can imagine.

Are you planning to work on some of my suggestions? If not I could also take a look at this, but i never used python before. Just that we do not both come up with an identical solution. I just wanted to ask before, so I dont try to work on a feature that is technically impossible.

**Comment by nims11** [comment]
Using id3 tags over folder names will be a better choice for artist based playlist generation (The user may choose between them).

I plan on implementing your suggestions one by one, including this one. If you want to work on it, I will more than glad. Implementing this feature won't be much tough, will just require adding few lines across the code.

**Comment by NicoHood** [comment]
But lets say you have a mixed playlist from someone who does not organize by artists. Then you have something like "party", "rock", "rap" etc. You better want to generate playlists per folder if you ask me. Also we then can apply this also for wav files very simple.

So I suggest two opens that will work at the same time:
* Make a playlist for each root folder (artists)
* Make a playlist for each subfolder (the last one in the chain)

With a structure like this:
`Ipod/ITunesfolder/Music/Artist/Album/*.mp3

And for people who do not use Artists/Album schema, they can just create playlists for all subfolders and have the same effect.

**Comment by NicoHood** [comment]
I suggest the following names:
```
--artist-playlist     Adds artist playlists inside /iPod_Control/Music/Artistfolders/ equal to --auto-playlist 1
--album-playlist      Adds album playlists inside /iPod_Control/Music/Artistfolders/Albumfolders/ equal to --auto-playlist 2
--auto-playlist N [M]   Adds recursive playlists for each folder inside /iPod_Control/Music/N/.../M M defaults to N
```

Maybe the description can be made clearer/shorter. What do you think of the idea?
`--auto-playlist 1 2` would generate artist and album playlists for example.

**Comment by NicoHood** [comment]
Here is my first attempt (relying on 1.2)

However I am still struggling with the path schema.
Also the depth in the folder counter has to be measured.
Proper input args have to be added of course. (current setting would be `--auto-playlist 1 infinite`)
Maybe we should also add a setting for infinite depth, as it is now. maybe zero would be an idea.

Suggestions always appreciated, as i am not that familiar with python. However I understand things better now. I just try to not change the whole script and rather apply minimal fixes with not bigger structure changes. The structure seems quite good.

```diff
diff --git a/shuffle.py b/shuffle.py
index 435c181..ce3c678 100755
--- a/shuffle.py
+++ b/shuffle.py
@@ -430,6 +430,28 @@ class Playlist(Record):
         listtracks = [ x for (_, x) in sorted(sorttracks) ]
         return listtracks
 
+    def populate_directory(self, playlistpath):
+        # Add all tracks inside the folder and its subfolders
+        #TODO
+        print "Playlist folder", playlistpath
+        sorttracks = []
+
+        for (dirpath, dirnames, filenames) in os.walk(playlistpath):
+            dirnames.sort()
+
+            for filename in sorted(filenames, key = lambda x: x.lower()):
+                fullPath = os.path.abspath(os.path.join(dirpath, filename))
+                relPath = fullPath[fullPath.index(playlistpath)+len(playlistpath)+1:].lower()
+                fullPath = os.path.abspath(os.path.join(playlistpath, relPath));
+                if os.path.splitext(filename)[1].lower() in (".mp3", ".m4a", ".m4b", ".m4p", ".aa", ".wav"):
+                    print "specialtrack", relPath
+                    # TODO get length and the rel path to the ipod root dir
+                    # example: iPod_Control/Music/Slipknot/All Hope Is Gone/11 - Snuff.mp3
+                    #sorttracks.append((num, filename))
+
+        listtracks = [ x for (_, x) in sorted(sorttracks) ]
+        return listtracks
+
     def remove_relatives(self, relative, filename):
         base = os.path.dirname(os.path.abspath(filename))
         if not os.path.exists(relative):
@@ -441,14 +463,22 @@ class Playlist(Record):
         return fullPath
 
     def populate(self, filename):
-        with open(filename, 'rb') as f:
-            data = f.readlines()
-
-        extension = os.path.splitext(filename)[1].lower()
-        if extension == '.pls':
-            self.listtracks = self.populate_pls(data)
-        elif extension == '.m3u':
-            self.listtracks = self.populate_m3u(data)
+        # Create a playlist of the folder and all subfolders
+        if os.path.isdir(filename):
+            print "playlist found"
+            self.listtracks = self.populate_directory(filename)
+
+        # Read the playlist file
+        else:
+            with open(filename, 'rb') as f:
+                data = f.readlines()
+
+            extension = os.path.splitext(filename)[1].lower()
+            if extension == '.pls':
+                self.listtracks = self.populate_pls(data)
+            elif extension == '.m3u':
+                self.listtracks = self.populate_m3u(data)
+
         # Ensure all paths are not relative to the playlist file
         for i in range(len(self.listtracks)):
             self.listtracks[i] = self.remove_relatives(self.listtracks[i], filename)
@@ -515,11 +545,13 @@ class Shuffler(object):
 
     def populate(self):
         self.tunessd = TunesSD(self)
+
         for (dirpath, dirnames, filenames) in os.walk(self.path):
             dirnames.sort()
             # Ignore the speakable directory and any hidden directories
             if "ipod_control/speakable" not in dirpath.lower() and "/." not in dirpath.lower():
                 for filename in sorted(filenames, key = lambda x: x.lower()):
+                    # TODO all this variables required?
                     fullPath = os.path.abspath(os.path.join(dirpath, filename))
                     relPath = fullPath[fullPath.index(self.path)+len(self.path)+1:].lower()
                     fullPath = os.path.abspath(os.path.join(self.path, relPath));
@@ -528,6 +560,13 @@ class Shuffler(object):
                     if os.path.splitext(filename)[1].lower() in (".pls", ".m3u"):
                         self.lists.append(os.path.abspath(os.path.join(dirpath, filename)))
 
+            # Create automatic playlists in music directory.
+            # Ignore the (music) root and any hidden directories.
+            #TODO check depth
+            if "ipod_control/music/" in dirpath.lower() and "/." not in dirpath.lower():
+                print "Auto Playlist Folder:", dirpath
+                self.lists.append(os.path.abspath(dirpath))
+
     def write_database(self):
         with open(os.path.join(self.base, "iPod_Control", "iTunes", "iTunesSD"), "wb") as f:
             f.write(self.tunessd.construct())
```

**Comment by NicoHood** [comment]
I got it working (not tested on hardware).
The real difficulty was to get the right path in the right variables with the correct lower/upper case.

The problem is, that for accessing the tracks the script uses the full path to the mp3. But upper/lower case is important here. The base path to the ipod (passed via command line) is upper AND lower case, while the path to the file and the file itself is lower. That makes it complex to construct correct filepaths. I do not know if this could be made simpler. "however it works now".

What has to be done is to add a depth checking and some command line parameters. At the moment the script just adds every folder as playlist.

Also if you have `/Artist/Album/files.mp` With a single Album it makes no sense to generate an Artist playlist (one of my TODO comments). Here the script should just search the subdirectory for further playlist folders. An Artist playlist just makes sense if you have multiple albums (or local files in the artist folder (which would be a bad file sorting though).

```diff
diff --git a/shuffle.py b/shuffle.py
index 274f0e0..0d42a55 100755
--- a/shuffle.py
+++ b/shuffle.py
@@ -450,6 +450,28 @@ class Playlist(Record):
         listtracks = [ x for (_, x) in sorted(sorttracks) ]
         return listtracks
 
+    def populate_directory(self, playlistpath):
+        # Add all tracks inside the folder and its subfolders recursively
+        listtracks = []
+        for (dirpath, dirnames, filenames) in os.walk(playlistpath):
+            dirnames.sort()
+
+            for filename in sorted(filenames, key = lambda x: x.lower()):
+                # Only add valid music files to playlist
+                if os.path.splitext(filename)[1].lower() in (".mp3", ".m4a", ".m4b", ".m4p", ".aa", ".wav"):
+                    # Reformat fullPath so that the basepath is lower/upper and the rest lower.
+                    # This is required to get the correct position (track index) inside Playlist.construct()
+                    # /media/username/USER'S IPOD/IPod_Control/Music/Artist/Album/Track.mp3
+                    fullPath = os.path.abspath(os.path.join(dirpath, filename))
+                    # /media/username/USER'S IPOD/
+                    basepath = self.base
+                    # ipod_control/music/artist/album/track.mp3
+                    ipodpath = self.path_to_ipod(fullPath)[1:].lower()
+                    # /media/username/USER'S IPOD/ipod_control/music/artist/album/track.mp3
+                    fullPath = os.path.abspath(os.path.join(basepath, ipodpath))
+                    listtracks.append(fullPath)
+        return listtracks
+
     def remove_relatives(self, relative, filename):
         base = os.path.dirname(os.path.abspath(filename))
         if not os.path.exists(relative):
@@ -461,17 +483,26 @@ class Playlist(Record):
         return fullPath
 
     def populate(self, filename):
-        with open(filename, 'rb') as f:
-            data = f.readlines()
-
-        extension = os.path.splitext(filename)[1].lower()
-        if extension == '.pls':
-            self.listtracks = self.populate_pls(data)
-        elif extension == '.m3u':
-            self.listtracks = self.populate_m3u(data)
-        # Ensure all paths are not relative to the playlist file
-        for i in range(len(self.listtracks)):
-            self.listtracks[i] = self.remove_relatives(self.listtracks[i], filename)
+        # Create a playlist of the folder and all subfolders
+        if os.path.isdir(filename):
+            self.listtracks = self.populate_directory(filename)
+
+        # Read the playlist file
+        else:
+            with open(filename, 'rb') as f:
+                data = f.readlines()
+
+            extension = os.path.splitext(filename)[1].lower()
+            if extension == '.pls':
+                self.listtracks = self.populate_pls(data)
+            elif extension == '.m3u':
+                self.listtracks = self.populate_m3u(data)
+            else:
+                print "Error" #TODO
+
+            # Ensure all paths are not relative to the playlist file
+            for i in range(len(self.listtracks)):
+                self.listtracks[i] = self.remove_relatives(self.listtracks[i], filename)
 
         # Handle the VoiceOverData
         text = os.path.splitext(os.path.basename(filename))[0]
@@ -535,11 +566,13 @@ class Shuffler(object):
 
     def populate(self):
         self.tunessd = TunesSD(self)
+
         for (dirpath, dirnames, filenames) in os.walk(self.path):
             dirnames.sort()
             # Ignore the speakable directory and any hidden directories
             if "ipod_control/speakable" not in dirpath.lower() and "/." not in dirpath.lower():
                 for filename in sorted(filenames, key = lambda x: x.lower()):
+                    # TODO all this variables required?
                     fullPath = os.path.abspath(os.path.join(dirpath, filename))
                     relPath = fullPath[fullPath.index(self.path)+len(self.path)+1:].lower()
                     fullPath = os.path.abspath(os.path.join(self.path, relPath));
@@ -548,6 +581,13 @@ class Shuffler(object):
                     if os.path.splitext(filename)[1].lower() in (".pls", ".m3u"):
                         self.lists.append(os.path.abspath(os.path.join(dirpath, filename)))
 
+            # Create automatic playlists in music directory.
+            # Ignore the (music) root and any hidden directories.
+            #TODO check depth
+            #TODO check if folder only contains a single folder -> skip
+            if "ipod_control/music/" in dirpath.lower() and "/." not in dirpath.lower():
+                self.lists.append(os.path.abspath(dirpath))
+
     def write_database(self):
         with open(os.path.join(self.base, "iPod_Control", "iTunes", "iTunesSD"), "wb") as f:
             f.write(self.tunessd.construct())
```

**Comment by NicoHood** [comment]
About the issue with folders containing a single subfolder:

```
Eminem
 -album1Eminem
 -album2Eminem
 -album3Eminem
David Guetta
 -album1guetta
```

Normally you'd generate a playlist with eminem (containing all eminem albums) and so for david guetta. However it makes no sense to have a david playlist as it is only a single album inside so, david and album1guetta would have the same content. There would be a difference if the david folder itself contain some music though.

I got a half working solution, I could finish it. However I thing this is too complex as then an error for an empty playlist would be generated. Its not a problem but doesnt look too nice. Avoiding this is more complex. Also I think it might just make sense to have just two playlist with the same content. Its just the same, but doesnt hurt. Via voiceover you have a david playlist and the same content for his single album. And if you ever add another album its just bigger next time.

So I intend to not implement this feature.

For documentation purposes here is my (not finished, disordered, just pasted) code:
```py
    def populate_directory(self, playlistpath, recursive):
        # Add all tracks inside the folder and its subfolders recursively
        #TODO check if folder only contains a single folder and no tracks -> skip
        # TODO this could be done with adding folder tracks, check if != 0 or subfolders > 1 (alos check if subfolders contain music)
        # TODO a recursive call to the function itself can add all subfolders
        # TODO (but requires a main (with only the check), to also add empty folders inside empty folders)
        listtracks = []
        for (dirpath, dirnames, filenames) in os.walk(playlistpath):
            dirnames.sort()

            for filename in sorted(filenames, key = lambda x: x.lower()):
                # Only add valid music files to playlist
                if os.path.splitext(filename)[1].lower() in (".mp3", ".m4a", ".m4b", ".m4p", ".aa", ".wav"):
                    # Reformat fullPath so that the basepath is lower/upper and the rest lower.
                    # This is required to get the correct position (track index) inside Playlist.construct()
                    # /media/username/USER'S IPOD/IPod_Control/Music/Artist/Album/Track.mp3
                    fullPath = os.path.abspath(os.path.join(dirpath, filename))
                    # /media/username/USER'S IPOD/
                    basepath = self.base
                    # ipod_control/music/artist/album/track.mp3
                    ipodpath = self.path_to_ipod(fullPath)[1:].lower()
                    # /media/username/USER'S IPOD/ipod_control/music/artist/album/track.mp3
                    fullPath = os.path.abspath(os.path.join(basepath, ipodpath))
                    listtracks.append(fullPath)
            if not recursive:
                break
        return listtracks

    def populate_directory_safe(self, playlistpath):
        # Add all tracks inside the folder and its subfolders recursively
        # Only add Playlist if directory or at least 2 subfolders contain music

        #if populate_directory(self, playlistpath, False) > 0
        # OR check all subfolders if >1 contain music
        #then
        # populate_directory(self, playlistpath, True)
        useDirectory = False
        listtracks = []
        listtracks += self.populate_directory(playlistpath, False)
        if len(listtracks):
            useDirectory = True

        for (dirpath, dirnames, filenames) in os.walk(playlistpath):
            dirnames.sort()

            for dirname in dirnames:
                print "dirname", dirname
                tracks = []
                # TODO firname needs to be merged with base
                tracks = self.populate_directory(dirname, True)
                if len(tracks):
                    print "yooooooooooooooo"
            break
        print useDirectory
        return listtracks
````

**Comment by NicoHood** [comment]
The fix is available here:
https://github.com/nims11/IPod-Shuffle-4g/compare/master...NicoHood:autoplaylist

I will also try to fix #26 and then prepare a PR

**Comment by NicoHood** [comment]
Fixed in #29
